core/trino-main/src/main/java/io/trino/operator/scalar/CharacterStringCasts.java

```diff
@@ -92,66 +88,67 @@ public static Slice charToVarcharCast(@LiteralParameter("x") Long x, @LiteralPar
     {
-        List<Integer> codePoints = new ArrayList<>(toCodePoints(slice));
+        IntList codePoints = toCodePoints(slice);
 
         // if Varchar(x) value length (including spaces) is greater than y, we can just truncate it
         if (codePoints.size() >= y) {
             // char(y) slice representation doesn't contain trailing spaces
-            codePoints = trimTrailing(codePoints, ' ');
-            List<Integer> codePointsTruncated = codePoints.stream()
-                    .limit(y)
-                    .collect(toImmutableList());
-            return codePointsToSliceUtf8(codePointsTruncated);
+            codePoints.size(Math.min((int) y, codePoints.size()));
+            trimTrailing(codePoints, ' ');
+            return codePointsToSliceUtf8(codePoints);
         }
 
         /*
          * Value length is smaller than same-represented char(y) value because input varchar has length lower than y.
          * We decrement last character in input (in fact, we decrement last non-zero character) and pad the value with
          * max code point up to y characters.
          */
-        codePoints = trimTrailing(codePoints, '\0');
+        trimTrailing(codePoints, '\0');
 
-        if (codePoints.isEmpty()) {
+        if (codePoints.size() == 0) {
             // No non-zero characters in input and input is shorter than y. Input value is smaller than any char(4) casted back to varchar, so we return the smallest char(4) possible
-            return codePointsToSliceUtf8(nCopies(toIntExact(y), (int) '\0'));
+            return Slices.allocate((int) y);
         }
 
-        codePoints = new ArrayList<>(codePoints);
         codePoints.set(codePoints.size() - 1, codePoints.get(codePoints.size() - 1) - 1);
-        codePoints.addAll(nCopies(toIntExact(y) - codePoints.size(), Character.MAX_CODE_POINT));
+        int toAdd = toIntExact(y) - codePoints.size();
+        for (int i = 0; i < toAdd; i++) {
+            codePoints.add(Character.MAX_CODE_POINT);
+        }
 
         verify(codePoints.get(codePoints.size() - 1) != ' '); // no trailing spaces to trim
 
         return codePointsToSliceUtf8(codePoints);
     }
 
-    private static List<Integer> trimTrailing(List<Integer> codePoints, int codePointToTrim)
+    private static void trimTrailing(IntList codePoints, int codePointToTrim)
     {
         int endIndex = codePoints.size();
         while (endIndex > 0 && codePoints.get(endIndex - 1) == codePointToTrim) {
             endIndex--;
         }
-        return ImmutableList.copyOf(codePoints.subList(0, endIndex));
+        codePoints.size(endIndex);
     }
 
-    private static List<Integer> toCodePoints(Slice slice)
+    private static IntList toCodePoints(Slice slice)
     {
-        ImmutableList.Builder<Integer> codePoints = ImmutableList.builder();
+        IntList codePoints = new IntArrayList(slice.length() / 4); //Smallest possible
```

Comment:
why not just `slice` and avoid resizing?

Also:
```
// Smallest possible
```

Reply:
Resizing is cheap and the memory is.. cheap as well. Most chars will be ASCII anyway so the resulting array will be on the order of magnitude of `slice.length` so I guess you're right.